### PR TITLE
More JEST 

### PR DESCRIPTION
**Pull request by revolunet**

### CODE REVIEW

**Comment by revolunet** [comment]
currently working with `--runInBand`

File: packages/api/routes/commentaires.js [hunk 1]
```diff
@@ -87,35 +87,28 @@ router.get(
...
-            .then(function(commentaires) {
-              res.status(200).json(commentaires);
-            });
-        })
-        .catch(function(error) {
-          console.log(error);
-          next(error);
-        });
-    });
+    const ti = await getTiByUserId(req.user.id);
```
Comment (douglasduteil) [inline packages/api/routes/commentaires.js]: We might want to try catch the would function and call `next(err)`

File: packages/api/routes/commentaires.js [hunk 2]
```diff
@@ -156,6 +149,15 @@ router.delete(
   async (req, res, next) => {
     // secu : ensure TI can write on this mandataire + add related test
     const ti = await getTiByUserId(req.user.id);
```
Comment (douglasduteil) [inline packages/api/routes/commentaires.js]: We might want to try catch the whole function and call `next(err)`

File: packages/api/routes/importation-excel.js [hunk 3]
```diff
@@ -16,12 +16,26 @@ router.post(
   typeRequired("individuel", "prepose"),
   async (req, res) => {
     const mandataire = await getMandataireByUserId(req.user.id);
```
Comment (douglasduteil) [inline packages/api/routes/importation-excel.js]: We might want to try catch the whole function and call `next(err)`

File: packages/api/__test__/e2e/auth/POST_login.test.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+//
+
+afterAll(async () => {
+  await knex.destroy();
+});
+
+test.each`
+  username       | password        | type            | url
+  ${"jeremy"}    | ${"johnson123"} | ${"individuel"} | ${"/mandataires"}
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/auth/POST_login.test.js]: url redirection was already on the snapshot no ?

**Comment by douglasduteil** [comment]
> currently working with `--runInBand`

which test require a sequential process ? you can use the global `const { knex } = global` to access to a temporary db

File: packages/api/__test__/e2e/auth/POST_login.test.js (hunk 4 above)
Comment (revolunet) [reply packages/api/__test__/e2e/auth/POST_login.test.js]: mmm not sure, looks like only the token was presetn ? https://github.com/SocialGouv/emjpm/blob/07ae0d585c733ceff9c3ffce0c8df2593916ec85/packages/api/__test__/e2e/auth/POST_login.test.js#L28

**Comment by revolunet** [comment]
most tests fail when run concurrently atm :/

```sh
Test Suites: 8 failed, 3 passed, 11 total
Tests:       29 failed, 44 passed, 73 total
Snapshots:   10 failed, 31 passed, 41 total
Time:        18.011s
```

File: packages/api/__test__/e2e/auth/POST_login.test.js (hunk 4 above)
Comment (douglasduteil) [reply packages/api/__test__/e2e/auth/POST_login.test.js]: Humm looks like its already there no ?

https://github.com/SocialGouv/emjpm/blob/07ae0d585c733ceff9c3ffce0c8df2593916ec85/packages/api/__test__/e2e/auth/__snapshots__/POST_login.test.js.snap#L33-L47

I don't think you need to specify the `url` as adding it or not is not changing the test snapshot :thinking:

Comment (revolunet) [reply packages/api/__test__/e2e/auth/POST_login.test.js]: okkkk got it thanks

File: packages/api/__test__/e2e/commentaires/DELETE_commentaires.test.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// strip the created_at date in the response
+const simplerComment = ({ created_at, ...props }) => props;
+
+test("should delete commentaire for given mandataire and return response", async () => {
+  const token = await getTokenByUserType("ti");
+  const response = await request(server)
+    .get("/api/v1/mandataires/1/commentaires")
+    .set("Authorization", "Bearer " + token);
+  const curComments = response.body;
+  console.log(
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/commentaires/DELETE_commentaires.test.js]: Is this console.log necessary ?

File: packages/api/__test__/e2e/commentaires/GET_commentaires.test.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+test("user cannot read other mandataires commentaires", async () => {
+  const token = await getTokenByUserType("ti");
+  const response = await request(server)
+    .get("/api/v1/mandataires/3/commentaires")
+    .set("Authorization", "Bearer " + token);
+
+  expect(response.body.length).toEqual(0);
+  expect(response.status).toBe(200);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/commentaires/GET_commentaires.test.js]: `user cannot read other mandataires commentaires` response status `200` is weird no ?
It should be `401 Unauthorized` or `404 Not Found` no ?

File: packages/api/__test__/e2e/commentaires/GET_commentaires.test.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+test("user cannot read other mandataires commentaires", async () => {
+  const token = await getTokenByUserType("ti");
+  const response = await request(server)
+    .get("/api/v1/mandataires/3/commentaires")
+    .set("Authorization", "Bearer " + token);
+
+  expect(response.body.length).toEqual(0);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/commentaires/GET_commentaires.test.js]: I would refer `inline snapshot` to be constant and explicit

Comment (revolunet) [reply packages/api/__test__/e2e/commentaires/GET_commentaires.test.js]: show me :)

File: packages/api/__test__/e2e/commentaires/GET_commentaires.test.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+test("ti cannot read other tis commentaires", async () => {
+  const token = await getTokenByUserType("ti");
+  const response = await request(server)
+    .get("/api/v1/mandataires/1/commentaires")
+    .set("Authorization", "Bearer " + token);
+
+  // there are 3 comments on mandataire_id=1 in the seeds, 2 from userTypes.ti
+  expect(response.body.length).toEqual(2);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/commentaires/GET_commentaires.test.js]: I would refer `inline snapshot` (if it's short) or a snapshot :)

File: packages/api/__test__/e2e/commentaires/GET_commentaires.test.js (hunk 6 above)
Comment (revolunet) [reply packages/api/__test__/e2e/commentaires/GET_commentaires.test.js]: fixed to 401

File: packages/api/__test__/e2e/inscription/POST_inscription.test.js [hunk 9]
```diff
@@ -113,9 +111,6 @@ test("should NOT register when username already exist", async () => {
       username: "jeremy"
     });
 
-  expect(response.body).toMatchSnapshot({
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/inscription/POST_inscription.test.js]: IMO an API should always return something inside its body. We should always snapshot bodies :camera_flash:

File: packages/api/__test__/e2e/mandataires/POST_mesures.test.js [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+  const currentMesures = await knex("mesures").where({
+    mandataire_id: 1,
+    status: "Mesure en cours"
+  });
+  const response = await request(server)
+    .post("/api/v1/mandataires/1/mesures")
+    .set("Authorization", "Bearer " + token)
+    .send(sampleMesure);
+
+  expect(response.body.length).toBe(currentMesures.length + 1);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/mandataires/POST_mesures.test.js]: This is implicitly tested by the snapshot below

File: packages/api/__test__/e2e/mandataires/PUT_mesures.test.js [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+  expect(newMesure.etablissement).toBe("DRUID");
+
+  expect(response.body).toMatchSnapshot();
+  expect(response.status).toBe(200);
+});
+
+test("mandataire should NOT PUT other mandataire mesure", async () => {
+  const token = await getTokenByUserType("mandataire");
+
+  const newMesureId = await knex("mesures")
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/mandataires/PUT_mesures.test.js]: You could use destructuring here by removing `.then(res => res[0]` and using `const [newMesureId]`

File: packages/api/__test__/e2e/mandataires/PUT_mesures.test.js [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    });
+
+  const newMesure = await knex("mesures")
+    .where({ id: newMesureId })
+    .first();
+
+  expect(newMesure.etablissement).toBe("ROOKIE");
+
+  expect(response.body).toMatchSnapshot();
+  expect(response.status).toBe(200);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/mandataires/PUT_mesures.test.js]: > mandataire should NOT PUT other mandataire mesure 

Return `200` is strange no ? :thinking:

File: packages/api/__test__/e2e/mandataires/PUT_mesures.test.js [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    });
+
+  const newMesure = await knex("mesures")
+    .where({ id: newMesureId })
+    .first();
+
+  expect(newMesure.etablissement).toBe("DRUID");
+
+  expect(response.body).toMatchSnapshot();
+  expect(response.status).toBe(200);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/mandataires/PUT_mesures.test.js]: If it's a creation we should return `201 Created` no ?

File: packages/api/__test__/e2e/mandataires/PUT_mesures.test.js [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    });
+
+  const newMesure = await knex("mesures")
+    .where({ id: newMesureId })
+    .first();
+
+  expect(newMesure.etablissement).toBe("ROOKIE");
+
+  expect(response.body).toMatchSnapshot();
+  expect(response.status).toBe(200);
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/mandataires/PUT_mesures.test.js]: > TI should NOT PUT other TI mesure

Return `200` is strange no ? :thinking:

File: packages/api/__test__/e2e/mandataires/__snapshots__/GET_mesures.test.js.snap [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    "residence": null,
+    "status": "Mesure en attente",
+    "ti_id": null,
+    "type": "preposes",
+    "ville": null,
+  },
+]
+`;
+
+exports[`should get list of mesures eteintes 1`] = `Array []`;
```
Comment (douglasduteil) [inline packages/api/__test__/e2e/mandataires/__snapshots__/GET_mesures.test.js.snap]: An empty array is it normal here ?

File: packages/api/db/queries/mandataires.js [hunk 16]
```diff
@@ -40,6 +40,7 @@ const updateDateMesureUpdate = id =>
 
 const getMandataire = filters =>
   knex("mandataires")
+    //.debug()
```
Comment (douglasduteil) [inline packages/api/db/queries/mandataires.js]: To remove ?

File: packages/api/routes/commentaires.js [hunk 17]
```diff
@@ -35,17 +36,25 @@ router.get(
...
+      }
+      const isAllowed = await isMandataireInTi(req.params.mandataireId, ti.id);
+      if (!isAllowed) {
+        throw createError.Unauthorized(`Mandataire not found in current TI`);
+      }
+      getAllCommentaires(req.params.mandataireId, ti.id)
+        .then(function(commentaires) {
+          res.status(200).json(commentaires);
+        })
+        .catch(function(error) {
```
Comment (douglasduteil) [inline packages/api/routes/commentaires.js]: This catch is unnecessary as it's just *next-ing* the error. Use using an await will auto throw it to the `try/catch`.

File: packages/api/routes/commentaires.js [hunk 18]
```diff
@@ -35,17 +36,25 @@ router.get(
...
+      const ti = await getTiByUserId(req.user.id);
+      if (!ti) {
+        throw createError.Unauthorized(`TI not found`);
+      }
+      const isAllowed = await isMandataireInTi(req.params.mandataireId, ti.id);
+      if (!isAllowed) {
+        throw createError.Unauthorized(`Mandataire not found in current TI`);
+      }
+      getAllCommentaires(req.params.mandataireId, ti.id)
+        .then(function(commentaires) {
```
Comment (douglasduteil) [inline packages/api/routes/commentaires.js]: No then needed here. 

```js
const commentaires = await getAllCommentaires(req.params.mandataireId, ti.id);
res.status(200).json(commentaires);
``

File: packages/api/routes/mandataireMesures.js [hunk 19]
```diff
@@ -127,31 +129,34 @@ router.put(
   "/:mandataireId/mesures/:mesureId",
   typeRequired("individuel", "prepose", "ti"),
   async (req, res, next) => {
+    if (Object.keys(req.body).length === 0) {
+      return res.status(200).json();
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: Why `PUT`-ing an empty body is returning a successful `200` ?
You mean https://httpstatuses.com/204 ?

File: packages/api/routes/mandataireMesures.js [hunk 20]
```diff
@@ -184,34 +189,46 @@ router.post(
...
-        .then(() => updateCountMesures(mandataire.id))
-        // todo : trigger/view
-        .catch(error => {
-          console.log(error);
-          next(error);
-        });
+    try {
+      if (Object.keys(req.body).length === 0) {
+        // return empty response when no data passed
+        res.status(200).json();
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: Why `PUT`-ing an empty body is returning a successful `200` ?
You mean https://httpstatuses.com/204 ?

File: packages/api/routes/mandataireMesures.js [hunk 21]
```diff
@@ -184,34 +189,46 @@ router.post(
...
+      };
+      if (!body.mandataire_id) {
+        throw createError.UnprocessableEntity("Mandataire not found");
+      }
+      if (req.user.type === "individuel" || req.user.type === "prepose") {
+        addMesure(body)
+          .then(() => updateCountMesures(body.mandataire_id))
+          .then(() => getMesuresEnCoursMandataire(body.mandataire_id))
+          .then(mesures => res.status(200).json(mesures))
+          .catch(next);
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: prefer using an await in front of `addMesure` : `await addMesure(body)`

File: packages/api/routes/mandataireMesures.js [hunk 22]
```diff
@@ -184,34 +189,46 @@ router.post(
...
+            ti.id
+          );
+          // TI cannot post for some other TI mandataire
+          if (!isAllowed) {
+            throw createError.Unauthorized(`Mandataire not found`);
+          }
+          addMesure({ ...body, ti_id: ti.id, cabinet: ti.cabinet })
+            .then(() => reservationEmail(ti, body))
+            .then(() => res.status(200).json({ success: true }))
+            .catch(next);
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: prefer using an await in front of `addMesure` : `await addMesure(body)`

File: packages/api/routes/mandataireMesures.js [hunk 23]
```diff
@@ -240,14 +257,28 @@ router.post(
...
+        if (!isAllowed) {
+          throw createError.Unauthorized(`Mandataire not found`);
+        }
+      }
+      addMesure({
+        ...req.body
+      })
+        .then(mesures => res.status(200).json(mesures))
+        .catch(error => {
+          next(error);
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: prefer using an await in front of `addMesure` : `await addMesure(body)`

File: packages/api/routes/mandataireMesures.js [hunk 24]
```diff
@@ -271,12 +302,19 @@ router.post(
...
+  typeRequired("individuel", "prepose"),
   async (req, res, next) => {
-    const mandataire = await getMandataireByUserId(req.user.id);
-    getMesuresEnCoursMandataire(mandataire.id)
-      .then(mesures => res.status(200).json(mesures))
-      .catch(error => next(error));
+    try {
+      const mandataire = await getMandataireByUserId(req.user.id);
+      if (!mandataire) {
+        throw createError.Unauthorized(`Mandataire not found`);
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: `Mandataire not found` so `404 Not Found` return code no ?

File: packages/api/routes/mandataireMesures.js [hunk 25]
```diff
@@ -271,12 +302,19 @@ router.post(
...
+      }
+      getMesuresEnCoursMandataire(mandataire.id)
+        .then(mesures => res.status(200).json(mesures))
+        .catch(error => next(error));
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: prefer using an await in front of `getMesuresEnCoursMandataire` : 

```js
const mesures = await getMesuresEnCoursMandataire(mandataire.id)
res.status(200).json(mesures)
```

File: packages/api/routes/mandataireMesures.js [hunk 26]
```diff
@@ -303,9 +341,16 @@ router.get(
   "/:mandataireId/mesuresForMaps",
   typeRequired("individuel", "prepose"),
   async (req, res, next) => {
-    const mandataire = await getMandataireByUserId(req.user.id);
-    const mesures = await getMesuresMap(mandataire.id);
-    res.status(200).json(mesures);
+    try {
+      const mandataire = await getMandataireByUserId(req.user.id);
+      if (!mandataire) {
+        throw createError.Unauthorized(`Mandataire not found`);
```
Comment (douglasduteil) [inline packages/api/routes/mandataireMesures.js]: `Mandataire not found` so `404 Not Found` return code no ?

File: packages/api/routes/mandataires.js [hunk 27]
```diff
@@ -106,13 +106,21 @@ const { getTiByUserId } = require("../db/queries/tis");
...
-  if (!mandataire) {
-    return next(new Error(401));
+router.get(
+  "/1",
+  typeRequired("individuel", "prepose"),
+  async (req, res, next) => {
+    try {
+      const mandataire = await getMandataireByUserId(req.user.id);
+      if (!mandataire) {
+        return next(new Error(401));
```
Comment (douglasduteil) [inline packages/api/routes/mandataires.js]: To replace by an `createError`.
Looks like a *404* to me :thinking:

File: packages/api/__test__/e2e/mandataires/GET_mesures.test.js [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+    .set("Authorization", "Bearer " + token);
+
+  expect(response.body).toMatchSnapshot();
+  expect(response.status).toBe(200);
+});
+
+test("should get list of mesures eteintes", async () => {
+  const token = await getTokenByUserType("mandataire");
+  const response = await request(server)
+    .get("/api/v1/mandataires/1/mesures/Eteinte")
```
[the remark this answers is not in the document]
Comment (revolunet) [reply packages/api/__test__/e2e/mandataires/GET_mesures.test.js]: will be fixed in #16

File: packages/api/__test__/e2e/mandataires/PUT_mesures.test.js (hunk 12 above)
Comment (revolunet) [reply packages/api/__test__/e2e/mandataires/PUT_mesures.test.js]: should be fixed in #17

File: packages/api/__test__/e2e/mandataires/PUT_mesures.test.js (hunk 13 above)
Comment (revolunet) [reply packages/api/__test__/e2e/mandataires/PUT_mesures.test.js]: should be fixed in #17

File: packages/api/routes/mandataireMesures.js (hunk 26 above)
Comment (revolunet) [reply packages/api/routes/mandataireMesures.js]: in fact its the case where someone try to interact with someone else data...

**Comment by revolunet** [comment]
```sh
-------------------------------|----------|----------|----------|----------|-------------------|
File                           |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-------------------------------|----------|----------|----------|----------|-------------------|
All files                      |    69.99 |    67.65 |    42.78 |    70.27 |                   |
 auth                          |    52.63 |    44.44 |    35.71 |    52.17 |                   |
  _helpers.js                  |    29.27 |    16.67 |    29.41 |    27.78 |... 78,79,82,83,87 |
  local.js                     |    88.46 |      100 |    66.67 |     87.5 |          33,56,57 |
  passport.js                  |    55.56 |      100 |       20 |    55.56 |        6,10,14,17 |
 email                         |    89.29 |       75 |    68.75 |       92 |                   |
  index.js                     |    92.86 |    83.33 |      100 |    92.86 |                17 |
  inscription.js               |      100 |      100 |      100 |      100 |                   |
  password-confirmation.js     |      100 |      100 |      100 |      100 |                   |
  password-reset.js            |      100 |      100 |      100 |      100 |                   |
  reservation-email.js         |    90.91 |    66.67 |       75 |    90.91 |                55 |
  validation.js                |    55.56 |      100 |        0 |    71.43 |             32,39 |
 routes                        |    70.42 |    71.76 |    41.26 |    70.57 |                   |
  admin.js                     |    54.55 |      100 |        0 |    54.55 |... 79,180,182,185 |
  auth.js                      |    87.93 |       75 |    88.24 |    87.72 |... 94,117,122,233 |
  commentaires.js              |    93.18 |    78.57 |      100 |    93.18 |        42,103,164 |
  doc.js                       |    70.59 |      100 |        0 |       80 |          47,48,49 |
  email.js                     |    71.43 |      100 |        0 |    71.43 |              7,13 |
  importation-excel.js         |    84.85 |       80 |     87.5 |    84.85 |    22,68,73,74,78 |
  index.js                     |      100 |      100 |      100 |      100 |                   |
  inscription.js               |    92.86 |    88.89 |    94.12 |    92.68 |       302,324,346 |
  mandataireMesures.js         |    91.11 |    79.41 |      100 |    91.11 |... 43,348,380,412 |
  mandataires.js               |    75.68 |    72.22 |    68.75 |    75.68 |... 39,440,443,445 |
  mandatairesEtablissements.js |    33.33 |      100 |        0 |    33.33 |... 86,187,188,189 |
  mesures.js                   |    24.49 |        0 |        0 |    24.49 |... 11,214,215,216 |
  serviceAntennes.js           |     37.5 |      100 |        0 |     37.5 |... 22,127,130,133 |
  tis.js                       |    34.38 |      100 |        0 |    34.38 |... 47,168,169,170 |
  users.js                     |    71.43 |      100 |        0 |    71.43 |             19,27 |
  usersTi.js                   |    54.55 |        0 |        0 |    54.55 |    18,19,21,23,26 |
-------------------------------|----------|----------|----------|----------|-------------------|
Snapshot Summary
 › 5 snapshots updated from 3 test suites.

Test Suites: 16 passed, 16 total
Tests:       166 passed, 166 total
Snapshots:   5 updated, 53 passed, 58 total
Time:        60.184s
```

**Pull request merged**